Fixed #32492 -- Added TrigramWordSimilarity() and TrigramWordDistance() on PostgreSQL.

C4ptainCrunch opened the pull request:
[Ticket #32492](https://code.djangoproject.com/ticket/32492)

This PR adds the `trigram_word_similar` lookup and updates the documentation with it.

I also wrote a test, but for it to be meaningful, i needed a longer string that the 16 chars available on `CharFieldModel.field` in the tests. I did increase the limit to 64 and updated the migration at the same time. Is this the right approach or should it create a new model or a new migration ? (As the test database is recreated almost each time, i guessed that modifying a migration is not a big deal but i could be wrong)
I ran the tests on PostgreSQL 13.4 on Mac with Python 3.9.1

I also added support for the `WORD_SIMILARITY()` SQL function (with `TrigramWordSimilarity()` ) and the `<->>` word similarity operator (with `TrigramWordDistance()`).

felixxm commented on docs/ref/contrib/postgres/lookups.txt:
```diff
@@ -26,6 +26,28 @@ The ``trigram_similar`` lookup can be used on
     ['<City: Middlesbrough>']
 
+.. fieldlookup:: trigram_word_similar
+
```
`versionadded:: 4.0` annotation is missing.

felixxm commented on docs/ref/contrib/postgres/search.txt:
```diff
@@ -326,3 +327,42 @@ Usage example::
     ...     distance=TrigramDistance('name', test),
     ... ).filter(distance__lte=0.7).order_by('distance')
     [<Author: Katy Stevens>, <Author: Stephen Keats>]
+
```
Chop blank line.

felixxm commented on docs/ref/contrib/postgres/search.txt:
```diff
@@ -326,3 +327,42 @@ Usage example::
+``TrigramWordSimilarity``
+---------------------
+
+.. class:: TrigramWordSimilarity(expression, string, **extra)
```
`versionadded:: 4.0` annotation is missing. Also, I'd move `TrigramWordSimilarity` docs below the `TrigramSimilarity`.

felixxm commented on docs/ref/contrib/postgres/search.txt:
```diff
@@ -326,3 +327,42 @@ Usage example::
+``TrigramWordDistance``
+-------------------
+
+.. class:: TrigramWordDistance(expression, string, **extra)
```
`versionadded:: 4.0` annotation is missing.

felixxm commented on docs/ref/contrib/postgres/search.txt:
```diff
@@ -326,3 +327,42 @@ Usage example::
     [<Author: Katy Stevens>, <Author: Stephen Keats>]
+
+
+``TrigramWordSimilarity``
+---------------------
```
```suggestion
``TrigramWordSimilarity``
-------------------------
```

felixxm commented on docs/ref/contrib/postgres/search.txt:
```diff
@@ -326,3 +327,42 @@ Usage example::
+    ... ).filter(similarity__gt=0.3).order_by('-similarity')
+    [<Author: Katy Stevens>]
+
+``TrigramWordDistance``
+-------------------
```
```suggestion
``TrigramWordDistance``
-----------------------
```

felixxm commented on tests/postgres_tests/test_trigram.py:
```diff
@@ -55,6 +56,35 @@ def test_trigram_similarity_alternate(self):
+            self.Model.objects.annotate(
+                word_distance=TrigramWordDistance('field', 'cat'),
+            ).filter(word_distance__lte=0.7).order_by('word_distance'),
+            [('Cat sat on mat.', 0)],
```
Can we add a test for non-zero distance?

felixxm commented on django/contrib/postgres/search.py:
```diff
@@ -300,3 +300,18 @@ class TrigramSimilarity(TrigramBase):
     arg_joiner = ' <-> '
+
+
+class TrigramWordSimilarity(Func):
+    function = 'WORD_SIMILARITY'
+    output_field = FloatField()
+
+    def __init__(self, expression, string, **extra):
+        if not hasattr(string, 'resolve_expression'):
+            string = Value(string)
+        super().__init__(string, expression, **extra)
```
Do we need to swap arguments? IMO we want to keep the same order as in `SIMILARITY()` calls.
```suggestion
class TrigramWordSimilarity(TrigramBase):
    function = 'WORD_SIMILARITY'
```
e.g.
- `TrigramWordSimilarity('Cat sat on mat.', 'cat')` should be equal to `0.30769232` instead of `1`
- `TrigramWordDistance('Cat sat on mat.', 'cat')` should be equal to `0.6923077` instead of `0`

felixxm replied:
Paolo, Can you take a look? (\cc @pauloxnet) :point_up:

C4ptainCrunch replied:
This is a good question. This thing is that the arguments of `SIMILARITY()` are commutative where for `WORD_SIMILARITY()` they are not.

The rationale for the argument swap is that i wanted to keep the column/field name on the left of the Django call. But most often when making a full-text search, the "long" string (`'Cat sat on mat.'`) is in the column where the "short" one (`'cat'`) is an argument so it will mostly be called like `WORD_SIMILARITY(a_column, 'the argument')` and in the Django call, i wanted to keep the column name as the first argument like in `TrigramSimilarity`.

I don't know what is best, writing the patch with the arguments in the other order is easy too, i'll follow your decision :)

felixxm replied:
>  .. i'll follow your decision :)

Just asking, I'm not an expert :shrug:. We can wait for the second opinion from Paolo.

felixxm replied:
@hannseman @jamesturk Can I ask for your opinion? :point_up:

hannseman replied:
I think it's probably for the best to try and keep the order of the arguments the same as in PostgreSQL. Feels like people will be cross-referencing against the PostgreSQL documentation when using this and swapping the arguments could be confusing.

felixxm replied:
As far I'm aware if we want to keep the same order as in PostgreSQL, it should be :thinking: 
```python
TrigramWordSimilarity(string, expression)
```
For example
```python
TrigramWordSimilarity('cat', 'Cat sat on mat.')
```

hannseman replied:
@felixxm yes that's what I'm thinking.

pauloxnet replied:
> Paolo, Can you take a look? (\cc @pauloxnet) point_up

Sorry, I totally missed the notification of this. I'll take a look

C4ptainCrunch replied:
I'd like to keep this moving so i still have a chance to get this in time for the feature freeze for 4.0 🤞 

I'll change this PR to have 
```python
TrigramWordSimilarity(string, expression)
```
And keep the same order in SQL.

felixxm commented on django/contrib/postgres/search.py:
```diff
@@ -300,3 +300,18 @@ class TrigramSimilarity(TrigramBase):
+
+class TrigramWordDistance(TrigramBase):
+    function = ''
+    arg_joiner = ' <->> '
```
:thinking: 
```suggestion
    arg_joiner = ' <<-> '
```
Paolo, Can you take a look? (\cc @pauloxnet) :point_up:

C4ptainCrunch replied:
This is for the same reason as the argument swap in `TrigramWordSimilarity`

hannseman replied:
I wonder if using the `<->>` operator would use any indexes 🤔

C4ptainCrunch replied:
I just made a quick test on PostgreSQL 12.5 and it seems the query planner is smart enough to use the index with both operators. So i think we should use the order that make the most sense (i would suggest have the same order as `WORD_SIMILARITY`)

<details>
<summary>Details of both queries and EXPLAIN</summary>

```sql
EXPLAIN ANALYSE
SELECT id, search <->> 'findme' as dist
FROM public.catalog_index
ORDER BY dist
LIMIT 10;
```


```sql
EXPLAIN ANALYSE
SELECT id, 'findme' <<-> search as dist
FROM public.catalog_index
ORDER BY dist
LIMIT 10;
```

Both queries give me the exact same plan:
```
Limit  (cost=0.28..9.69 rows=10 width=20) (actual time=5.884..7.119 rows=10 loops=1)
  ->  Index Scan using catalog_index_search_gist_idx on catalog_index  (cost=0.28..11300.31 rows=12002 width=20) (actual time=5.883..7.115 rows=10 loops=1)
        Order By: (search <->> 'findme'::text)
Planning Time: 0.105 ms
Execution Time: 7.147 ms
```

</details>

felixxm reviewed (commented):
@C4ptainCrunch Thanks for this patch :+1: I left comments. Please add also release notes (`4.0.txt`).

hannseman commented on django/contrib/postgres/lookups.py:
```diff
@@ -58,3 +58,8 @@ def process_lhs(self, qn, connection):
+
+class TrigramWordSimilar(PostgresOperatorLookup):
+    lookup_name = 'trigram_word_similar'
+    postgres_operator = '%%>'
```
```suggestion
    postgres_operator = '<%%'
```

I might be missing something but don't we want to test the word similarity of an input against a column? Like this:

```sql
WHERE col <% 'input'
```

C4ptainCrunch commented:
Hello again 👋 

Thank you for your comment :)

I've implemented the fixed you asked for and rebased my branch.

Notable changes are:
- The order of the arguments in `TrigramWordSimilarity()` to keep the same order as in SQL and `TrigramWordDistance()` to so that both have the same order.
- A new `TrigramSwapedBase(string, expression)` class that takes the arguments in the other order as `TrigramBase(expression, string)`. I'm not extremely proud of the naming though
- Added release notes
- Changed a test so we don't check for a `0` distance

hannseman commented:
@C4ptainCrunch uncheck "Needs documentation" and "Patch needs improvement" in the ticket https://code.djangoproject.com/ticket/32492. It'll make the ticket appear in the review queue.

felixxm commented:
@C4ptainCrunch Thanks for updates :+1:  Welcome aboard :boat:

I pushed small edits.

pauloxnet commented on django/contrib/postgres/search.py:
```diff
@@ -293,10 +293,28 @@ def __init__(self, expression, string, **extra):
     arg_joiner = ' <-> '
+
+
+class TrigramWordDistance(TrigramWordBase):
+    function = ''
+    arg_joiner = ' <<-> '
+
+
+class TrigramWordSimilarity(TrigramWordBase):
+    function = 'WORD_SIMILARITY'
```
I would use the same order as the previous two classes (`TrigramSimilarity` and `TrigramDistance`):

```suggestion
class TrigramWordSimilarity(TrigramWordBase):
    function = 'WORD_SIMILARITY'


class TrigramWordDistance(TrigramWordBase):
    function = ''
    arg_joiner = ' <<-> '
```

felixxm replied:
@pauloxnet Ahh, sorry, just merged.

pauloxnet commented:
I realized you just merged, sorry, I was too slow to review, but I only found time today. I had entered a negligible suggestion, but everything else is fine.

felixxm commented:
@pauloxnet Thanks for checking :+1: The feature freeze is on Monday that's why I didn't wait longer.

pauloxnet commented:
I will try to be faster in my next reviews.
However, I did not find anything wrong with my review.
Great job @C4ptainCrunch @hannseman @felixxm :clap:
